File: solutions/haskell/clock/1/src/Clock.hs
-- import Text.Printf

module Clock (addDelta, fromHourMin, toString) where

data Clock = Clock
  { hours ::  Int,
    minutes :: Int
  }
  deriving Eq

fromHourMin :: Int -> Int -> Clock
fromHourMin hours minutes = Clock ((hours + (minutes `div` 60)) `mod` 24) (minutes `mod` 60)

toString :: Clock -> String
--toString (Clock h m) = printf "%02d:%02d" h m
toString (Clock hour minute) = (padZero hour) ++ ":" ++ (padZero minute)
  where
    padZero t 
      | t < 10 = '0' : (show t)
      | otherwise = show t

addDelta :: Int -> Int -> Clock -> Clock
addDelta hours minutes (Clock h m) = fromHourMin (hours + h) (minutes + m)

File: solutions/ruby/clock/1/clock.rb
class Clock
  attr_reader :hour, :minute
  def initialize(hour: 0 ,minute: 0)
    @minute = minute % 60
    @hour = (hour + (minute / 60)) % 24
  end

  def to_s()
    sprintf("%.2d:%.2d", @hour, @minute)
  end

  def +(clock)
    Clock.new(hour: @hour + clock.hour, minute: @minute + clock.minute)
  end

  def -(clock)
    Clock.new(hour: @hour - clock.hour, minute: @minute - clock.minute)
  end

  def ==(clock)
    return @hour == clock.hour && @minute == clock.minute
  end
end
